### PR TITLE
new comment functionality on articles

### PR DESCRIPTION
**Pull request by jonathondilworth**
* Added UserContext to be used across components that require any user details (e.g. username)
* Added a CommentForm
* CommentForm tracks user input, validates user input & submits a POST to save a comment
* Implemented isLoading pattern to stop multiple POST requests whilst one submission is processing
* Form cannot be submitted until inputs are validated
* Submission triggers a re-render within a subling component, by sharing ‘newComment’ state across via a parent

### CODE REVIEW

File: src/components/CommentForm.jsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        event.preventDefault();
+        if (!validateNewComment(newCommentInput)) {
+            alert("You cannot submit an empty comment");
+            return;
+        }
+        setIsLoading(true);
+        postNewComment(article.article_id, currentUser, newCommentInput)
+        .then((comment) => {
+            alert("New Comment Added!");
+            // triggers a re-render in ArticleComments.jsx
```
Comment (nc-poppy-m) [inline src/components/CommentForm.jsx:37]: I like the concept! However from a UX perspective, the rerender of the component makes everything feel a little 'jerky'. If your API returns the posted comment, I would probably recommend just disabling the form until it's posted, getting the posted comment back from the API and setting the comments in state with the posted one on top

File: src/components/CommentForm.jsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            setNewCommentInput("");
+            setIsLoading(false);
+        });
+    }
+
+    return isLoading ? <Loading /> : (
+        <form onSubmit={handleSubmit}>
+            <label htmlFor='new-comment'>
+            <input type="text" id="new-comment" value={newCommentInput} onChange={handleNewCommentChange} />
+            {newCommentInputError ? <span style={{"color": "red", "fontSize": "small"}}>{newCommentInputError}</span> : ''}
```
Comment (nc-poppy-m) [inline src/components/CommentForm.jsx:51]: It's good you're trying this! However, I couldn't work out how to trigger this if I submitted an invalid new comment

**Review by nc-poppy-m — COMMENTED** [review]
Great work on this! You've really thought these points through. I would reconsider the whole rerender at some stage though if possible, just because it makes it a little jerk-y form a UX perspective.

File: src/components/CommentForm.jsx (hunk 2 above)
Comment (jonathondilworth) [reply src/components/CommentForm.jsx:51]: If you type the comment in and then delete it (remove all characters from the input), it should trigger as it's just doing a 'required field' check - unless I've broken it! I'll have a look at it anyway, thanks!

**Pull request merged**